fix(processing): Filter kwargs in ProcessorMixin call to prevent Type…

CodersAcademy006 opened the pull request:
This PR fixes an issue where multimodal processors could raise a `TypeError` when called with arguments specific to one modality. The `ProcessorMixin.__call__` method was passing all keyword arguments to each sub-processor (tokenizer, feature extractor, etc.), causing a crash if a processor received an argument it didn't recognize (e.g., `EncodecFeatureExtractor` receiving `pad_to_multiple_of` from the tokenizer).

This fix resolves the problem by filtering the keyword arguments before they are passed to each sub-processor. It uses Python's `inspect` module to get the valid parameters for each processor's `__call__` method and only passes the arguments that match. This prevents `TypeError`s and makes the processor logic more robust.

**Fixes #41598**

Rocketknight1 commented:
cc @zucchini-nlp! I'm not sure about this because some of the submodules may just have `**kwargs`

zucchini-nlp commented on src/transformers/processing_utils.py:
```diff
@@ -626,7 +627,15 @@ def __call__(
             attribute = getattr(self, attribute_name, None)
             input_data, input_kwargs = attribute_to_kwargs[attribute_name]
             if input_data is not None and attribute is not None:
-                attribute_output = attribute(input_data, **kwargs[input_kwargs])
```
in this line `kwargs` are filtered already and we are obtaining `kwargs[input_kwargs]`. That should work for all models if if there's a TypeError anywhere, I'd suggest to see what has gone wrong with specific model

Commit:
fix(processing): filter modality kwargs for sub-processors with explicit signatures

`_merge_kwargs` routes kwargs into per-modality buckets based on TypedDict
membership. Several kwargs (e.g. `pad_to_multiple_of`) are declared in more
than one modality TypedDict (TextKwargs and AudioKwargs), so they land in
every matching bucket. When a concrete sub-processor has an explicit `__call__`
signature without a **kwargs catch-all it then receives unexpected keyword
arguments and raises a TypeError.

Fix: before calling each sub-processor, inspect its `__call__` signature.
- If a VAR_KEYWORD parameter exists the processor already handles unknown keys
  itself, so all modality kwargs are forwarded unchanged.
- Otherwise, filter modality kwargs to the set of params the sub-processor
  explicitly accepts.

Adds two regression tests in ProcessorMixinKwargsFilterTest that reproduce
the failure with a strict-signature feature extractor and confirm the
**kwargs path continues to work unchanged.

Fixes #41598

Qodo-Free-For-OSS commented:
Hi, `ProcessorMixin.__call__` unconditionally calls `inspect.signature(attribute.__call__)`, which can raise `TypeError`/`ValueError` for some callable implementations and would crash processing before invoking the sub-processor.

**Severity:** remediation recommended | **Category:** reliability

**How to fix:** Add signature() fallback path

**Agent prompt to fix** - you can give this to your LLM of choice:

> ### Issue description
> `ProcessorMixin.__call__` now relies on `inspect.signature(attribute.__call__)` to decide whether to filter kwargs. If `inspect.signature()` raises for a particular callable implementation, processing crashes before the sub-processor is called.
> 
> ### Issue Context
> Some sub-processors may be dynamically loaded or implemented in ways that are not inspectable.
> 
> ### Fix Focus Areas
> - src/transformers/processing_utils.py[689-703]
> 
> ### Suggested approach
> - Wrap `inspect.signature(attribute.__call__)` in `try/except (TypeError, ValueError)`.
> - On exception, fall back to a safe behavior. Recommended fallback:
>   - Attempt calling the sub-processor with full `modality_kwargs` (preserves current behavior for processors that tolerate extras via internal handling).
>   - If you want to preserve the original goal of avoiding TypeError, alternatively fall back to filtering using a conservative allowlist (e.g., empty filter) only if that won’t break required args.
> - Add/adjust a unit test that simulates an uninspectable `__call__` (e.g., a callable object whose `__call__` is a built-in or raises when inspected) and assert the processor call does not crash.

---
*Found by [Qodo](https://qodo.ai) code review*
